### ISSUE TITLE
GeoTIFF support

### ISSUE THREAD

**Issue by wieringen** [valhalla/valhalla#3695]
For the last couple of weeks, I've been experimenting with Valhalla to plot scenic bicycle routes, and so far, it's going pretty well. Ideally, I would like to provide elevation information for every route since this can be important for cyclists. I've been reading up on elevation data formats and the corresponding Valhalla documentation and enabled this feature for my instance. The experience has been great; only one flag to set, and it just works! The only problem I have is with the returned data; the resolution (~30m) is not great. 

After some research, I found that the Dutch government offers a high-resolution dataset for free. They provide a WMS service (https://geodata.nationaalgeoregister.nl/ahn3/wms?service=WMS&request=GetCapabilities) and 
offer direct download links to tiles (https://app.pdok.nl/ahn3-downloadpage).

My first (naive) thought was maybe I can download these tiles with a script and upload them to a storage bucket in the cloud, which I can reference in the "elevation_url" field inside valhalla_build_config. I quickly realized this wasn't feasible because the formats and projections differ. According to the documentation, Valhalla's elevation tiles are in the SKADI format, which is compressed SRTMGL1 using the EPSG:4326 coordinate system. The downloadable tiles (DTS, 0.5m) provided by the Dutch government are in GeoTIFF format; I used GDAL to get some additional information: 

```
Driver: GTiff/GeoTIFF
Files: M_27BN2.TIF (https://download.pdok.nl/rws/ahn3/v1_0/05m_dtm/M_27BN2.ZIP)
Size is 10000, 12500
Coordinate System is:
PROJCRS["Amersfoort / RD New",
    BASEGEOGCRS["Amersfoort",
        DATUM["Amersfoort",
            ELLIPSOID["Bessel 1841",6377397.155,299.1528128,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4289]],
    CONVERSION["RD New",
        METHOD["Oblique Stereographic",
            ID["EPSG",9809]],
        PARAMETER["Latitude of natural origin",52.1561605555556,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",5.38763888888889,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9999079,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",155000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",463000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["easting (X)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["northing (Y)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["Engineering survey, topographic mapping."],
        AREA["Netherlands - onshore, including Waddenzee, Dutch Wadden Islands and 12-mile offshore coastal zone."],
        BBOX[50.75,3.2,53.7,7.22]],
    ID["EPSG",28992]]
Data axis to CRS axis mapping: 1,2
Origin = (195000.000000000000000,500000.000000000000000)
Pixel Size = (0.500000000000000,-0.500000000000000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  INTERLEAVE=BAND
Corner Coordinates:
Upper Left  (  195000.000,  500000.000) (  5d58'35.79"E, 52d29'14.21"N)
Lower Left  (  195000.000,  493750.000) (  5d58'33.10"E, 52d25'51.98"N)
Upper Right (  200000.000,  500000.000) (  6d 3' 0.81"E, 52d29'12.81"N)
Lower Right (  200000.000,  493750.000) (  6d 2'57.78"E, 52d25'50.58"N)
Center      (  197500.000,  496875.000) (  6d 0'46.87"E, 52d27'32.41"N)
Band 1 Block=256x256 Type=Float32, ColorInterp=Gray
```

My second thought was maybe I can convert these tiles to the SRTMGL1/SKADI format using GDAL, but that doesn't seem possible.

```
gdal_translate -of SRTMHGT M_27BN2.TIF M_27BN2.hgt

Input file size is 10000, 12500
Warning 1: The source projection coordinate system is PROJCS["Amersfoort / R … ["EPSG","28992"]]. Only WGS 84 is supported. The SRTMHGT driver will generate a file as if the source was WGS 84 projection coordinate system.
Warning 1: The corner coordinates of the source are not properly aligned on plain latitude/longitude boundaries.
ERROR 1: Image dimensions should be 1201x1201, 3601x3601 or 1801x3601.
```

Looking at the output of GDAL, it seems the SRTMHGT format doesn't support a higher resolution than ~30m. Is this correct? How hard would it be to add support to Valhalla for a more generic format like GeoTIFF if it can't?

**Comment by kevinkreiser** [comment]
@wieringen i cant help but make this joke... a simple code change should fix you right up:

```c++
// pretty flat in the netherlands (its in the name!)
if (node->admin() == "NL")
  return 0;
```

your first thought was correct, we cant support any old elevation formatted data (at least not yet!) and yes currently we only have support for SRTMGL1 which is a 1 degree grid (ie unprojected 4326) with ~30m resolution pixels as you mentioned.

your second thought about converting it should actually work. you would need to use `gdalwarp` though so you could both reproject it to 4326 AND convert it to hgt format.

your final point is a good one. you are right, the srtm format does not support any higher resolution so even if you have awesome data in the netherlands it will get quantized to 30m when you convert it to hgt. it might improve your use case but as you mentioned 30m is pretty coarse. i will say this though, the reason why we use this format is due to the fact that it is an extremely simple one, its very easy to memory map which makes it very easy to serve the data in an api. that said i completely understand the desire for better data. to that i would say a couple things.

1. yes we could suport reading from a different format with higher or even variable resolution indeed by using gdals c++ api. i could see a mode where you give valhalla a gdal vrt and it simply uses that to get height values letting gdal do the hard work
2. this would probably never fly for real time use-cases like we support now where you can send us any old polyline and we can give you back elevation along it
3. we could come off of that use-case though and sequester the data to just along edges in the graph actually storing the data with the edge (in edge info). i know several others have done this with valhalla already and i could envision a scheme in which the resolution of this data could be set at tile build time. so if your application only needs 30m we store that and if you need 10m you store that, etc.

so yeah those would be some of the problems we'd need to solve to get you where you want to go. would you be interested in working on those. i have to say i would be but i cannot commit to it at this point in time (i have like 3 or 4 other in progress projects here plus my regular job haha)

**Comment by dnesbitt61** [comment]
@kevinkreiser - Is there a way to get this into the elevation set Valhalla uses in valhalla_get_elevation (e.g. skadi elevation tiles in AWS S3)?

**Comment by kevinkreiser** [comment]
@dnesbitt61 i didnt quite understand what you meant by that. when you say "get this into" are you saying get this NL data into the existing elevation tileset on aws?

if so, technically it is possible:

1. gdalwarp this nl data to 4326
2. download the existing tiles that cover this area of the world
3. make a vrt with both sets of data putting this new nl data last in the list (so that it wins where it has coverage)
4. use gdal translate to cut out the srtm tiles over this area from the vrt
5. upload those tiles over top of what is already there in s3

this ignores the highly likely possibility that these data sources will disagree where they butt up next to each other. in such regions we will likely have a slight step up or down where we can see the data mask around the new nl data.

**Comment by dnesbitt61** [comment]
yes, that is what I meant. the mismatch issue seems like a stopper. Not sure how one would handle that.

**Comment by kevinkreiser** [comment]
you need to do a linear combination of the two data sources (using alpha blending) based on the distance of a pixel from the seam. its totally doable but i dont know off the top of my head the off the shelf way to do it (though im sure gdal et al has something like this already since its a hugely common problem)

**Comment by nilsnolde** [comment]
Wasn’t this the project to compile all those different data sources? https://github.com/tilezen/joerd/blob/master/docs/formats.md

Only on mobile right now, but IIRC it supports more formats and does the heavy lifting via gdal bindings. Not very much maintained😅but I was able to run it years ago.

**Comment by kevinkreiser** [comment]
yes it was, and it originally lived in valhalla the problem though was that it just didnt handle at all the inconsistencies at the boundaries of data sources because when i first made it, there were only like 4 data sources. every time you add one you get more and more seams to deal with, i think jörd which is like skadi v2 already has lots of data sets in there so probably lots of seams but they also switch some out based on what zoom level is being targetted. there was then a v3 of this project called marble cutter.

**Comment by wieringen** [comment]
> @wieringen i cant help but make this joke... a simple code change should fix you right up:
Haha fair enough

> 1. yes we could support reading from a different format with higher or even variable resolution indeed by using gdals c++ api. i could see a mode where you give valhalla a gdal vrt and it simply uses that to get height values letting gdal do the hard work
Which format, CRS, tile size, and tiling structure would be a good starting point? GDAL VRT looks pretty cool. If I understand correctly, with VRT, it's possible to overlay your tiles with a virtual dataset. Afterward, it's possible to query the VRT file and GDAL will take care of efficiently loading/swapping the underlying tiles. Is that correct?

Is this what you had in mind for implementation?
- Call GDALOpen() to open the VRT file and get the dataset.
- Use GetGeoTransform() to fetch the coefficients need for transforming between raster space and projection coordinates.
- Get the elevation raster band with GetRasterBand() 
- Use RasterIO() to load the area corresponding to a Valhalla routing tile to memory.
- Get elevation for an edge from raster data and store it in edge info.

> so yeah those would be some of the problems we'd need to solve to get you where you want to go. would you be interested in working on those.
I'm not that familiar yet with the code base (or the subject matter) that I will be able to take up a feature of this magnitude. However, I plan to experiment with the techniques you just mentioned to get a better grasp.

**Comment by kevinkreiser** [comment]
@wieringen yeah it would be something like that. heres an example: https://gis.stackexchange.com/questions/393549/retrieve-value-by-geographic-coordinate-from-geotiff-with-gdal-c
